package-build.el

```diff
@@ -166,55 +192,27 @@ function for access to this function")
     (concat (format-time-string "%Y%m%d." time)
             (format "%d" (or (string-to-number (format-time-string "%H%M" time)) 0)))))
 
-(defun package-build--string-match-all (regex str &rest groups)
-  "Find every match for `REGEX' within `STR'.
-Return a list containing the full match string and match for
-groups `GROUPS'.  The return list is of the form
-   ((FULL GROUP1 GROUP2 ...) ...)
-where FULL is the complete regexp match and
-GROUP1, GROUP2, ... are the regex groups specified by the
-`GROUPS' argument.  If `GROUPS' is nil then FULL and GROUP1 will
-be identical."
-  (let (result
-        (pos 0)
-        (groups (or groups '(0))))
-    (while (string-match regex str pos)
-      (push (cons (match-string 0 str) (mapcar
-                                        (lambda (group)
-                                          (match-string group str))
-                                        groups))
-            result)
-      (setq pos (match-end 0)))
-    result))
-
 (defun package-build--find-parse-time (regex &optional bound)
   "Find REGEX in current buffer and format as a time version, optionally looking only as far as BOUND."
   (package-build--parse-time (progn (re-search-backward regex bound)
                         (match-string-no-properties 1))))
 
-(defun package-build--valid-version-string (str)
-  "Return true if STR is a valid version, otherwise return nil."
-  (ignore-errors (version-to-list str)))
-
-(defun package-build--find-tag-version-newest (regex &optional bound &rest additional-groups)
-  "Find the newest version matching REGEX after point, maybe stopping at BOUND.
-The first capture group 1 is examined, together with any ADDITIONAL-GROUPS."
-  (let* ((text (buffer-substring-no-properties
-                (or bound (point-min)) (point)))
-         (tags (cl-remove-if-not
-                (lambda (tag-version)
-                  (package-build--valid-version-string (cadr tag-version)))
-                (apply 'package-build--string-match-all regex text 1 additional-groups))))
-    (car (nreverse (sort tags (lambda (v1 v2)
-                                (version< (cadr v1) (cadr v2))))))))
-
-(defun package-build--find-parse-time-latest (regex &optional bound)
-  "Find the latest timestamp matching REGEX, optionally looking only as far as BOUND."
-  (let* ((text (buffer-substring-no-properties
-                (or bound (point-min)) (point)))
-         (times (mapcar 'package-build--parse-time
-                        (mapcar 'cadr (package-build--string-match-all regex text 1)))))
-    (car (nreverse (sort times 'string<)))))
+(defun package-build--find-version-newest (regex &optional bound)
```

Comment:
ok, I see my confusion here. alright. so here is my curiosity; can we change this function so that it returns a tuple `('tag name' (0 1 2))` which is essentially `(<original string> <version parsed out of string>)`.

Because *then* we don't need to worry about re-parsing in the `package-build-package` function. it should also simplify things.

and it's not hard to change. Anything that uses this function (e.g., cvs and svn) checkouts, can just grab the `cadr`? What do you think?

i think this is close enough that if you don't want to change it up I am willing to patch in and fix up, but I'm more than happy to have your help  :)